### PR TITLE
Use libsonata to resolve node sets

### PR DESCRIPTION
**Pull request by joni-herttuainen**

### CODE REVIEW

File: bluepysnap/simulation.py
Comment (joni-herttuainen) [inline bluepysnap/simulation.py:126]: Frankly, I am not sure if this should return the `NodeSets` object or the `NodeSets.content`. 

If the use case is resolving `Simulation`'s node sets, then `NodeSets`, if it's just for information, then content, I quess.

File: bluepysnap/node_sets.py
Comment (edasubert) [inline bluepysnap/node_sets.py:45]: I am not sure if `resolve_ids` is an appropriate name, I would expect to call `resolve` to resolve internally and then `get` to get the actual value; so I think in this case calling it `get_ids` straight away might be a good choice

Comment (joni-herttuainen) [reply bluepysnap/node_sets.py:45]: yeah, valid point. `get_ids` or `ids` is better

File: bluepysnap/nodes/node_population.py
Comment (GianlucaFicarelli) [reply bluepysnap/nodes/node_population.py:281]: Very good that the times are better, but I'm surprised that querying on Layer23Inhibitory is taking 1 second.

I did a quick check, and I see that most of the time is spent in doing 9 times (one for each mtype) something like:
```python
%time np.in1d(cells['mtype'], 'L23_SBC')
```
that takes around 110ms:
```
CPU times: user 94.2 ms, sys: 17.5 ms, total: 112 ms
Wall time: 111 ms
array([False, False, False, ..., False, False, False])
```
and it's not optimal because:
- mtype is a category, but using np.in1d requires a conversion to object (35ms), and then looking through strings instead of numbers (65ms)
- it should be called only once, passing all the mtypes as a list

For comparison, this calls 
```python
 %time cells['mtype'].isin(['L23_SBC', 'L23_BTC', 'L23_NGC','L23_MC','L23_NBC','L23_LBC','L23_BP','L23_CHC', 'L23_DBC']).to_numpy()
```
would require only 9ms to check all the 9 mtypes at the same time:
```
CPU times: user 9.02 ms, sys: 3 µs, total: 9.02 ms
Wall time: 8.44 ms
array([False, False, False, ..., False, False, False])
```

However, these considerations are unrelated to the MR. I can see if the np.in1d can be safely changed to make the query function more performant, and maybe we can repeat the comparison.

Comment (joni-herttuainen) [reply bluepysnap/nodes/node_population.py:281]: Thanks for profiling. 

> I did a quick check, and I see that most of the time is spent in doing 9 times (one for each mtype) something like:
> ```python
> %time np.in1d(cells['mtype'], 'L23_SBC')
> ```

I am not surprised. I guess there has not been a reason to optimize the queries. And, to be fair, 1s for a query is still okayish provided how much data it filters. Obviously, if we can easily reduce it to tenth of that, we should.

> I can see if the np.in1d can be safely changed to make the query function more performant, and maybe we can repeat the comparison.

Am I assuming correctly that you would rather keep the node set resolution in SNAP?

Comment (GianlucaFicarelli) [reply bluepysnap/nodes/node_population.py:281]: > Am I assuming correctly that you would rather keep the node set resolution in SNAP?

Nope, but it can be worth to improve the query function because it's used also when `get()` is called with generic queries

Comment (joni-herttuainen) [reply bluepysnap/nodes/node_population.py:281]: Ah, of course! Makes sense.

File: tests/data/node_sets.json
Comment (mgeplf) [reply tests/data/node_sets.json:10]: FYI: I should be closing that soon; we've cleared up the expected semantics.

File: bluepysnap/edges/edges.py
Comment (joni-herttuainen) [inline bluepysnap/edges/edges.py:49]: How (if at all) can we handle this without disabling `arguments-differ`?

File: bluepysnap/frame_report.py
Comment (joni-herttuainen) [inline bluepysnap/frame_report.py:243]: Still open to discussion if this should just return the node set name.

Usage would then be:
```python
sim.circuit.get(report.node_set, node_sets=sim.node_sets)
```

File: bluepysnap/network.py
Comment (joni-herttuainen) [inline bluepysnap/network.py:151]: I can try to get this with less locals in this PR, or we can review it when we make change the get in NSETM-2186.

**Review by joni-herttuainen — COMMENTED** [review]
@GianlucaFicarelli  Tell me what you think. If this looks somewhat ok, I'll update the docs and add changelog entries. Should we bump this up a major version bearing in mind we will also change `{nodes,edges}.get`?

File: bluepysnap/frame_report.py
Comment (GianlucaFicarelli) [reply bluepysnap/frame_report.py:243]: From what I see, `cells` is optional at https://sonata-extension.readthedocs.io/en/latest/sonata_simulation.html#reports

> Specify which node_set to report, default is the simulation “node_set”.

I suppose that returning the name should be enough, but it's better to wait for other opinions.

File: bluepysnap/edges/edges.py
Comment (GianlucaFicarelli) [reply bluepysnap/edges/edges.py:49]: Still for discussion, I was also considering another option that wouldn't require to change the signature of ids() and get().
It would be possible to define a class `NodeSet` that can be returned by `node_sets[node_set_name]`, and that can be passed as `group` to ids() and get().

So, instead of doing
```python
sim.circuit.get(name, node_sets=sim.node_sets)
```
you can override the circuit node_sets for example in this way:
```python
sim.circuit.get(sim.node_sets[name])
```
then get() can internally use the NodeSet to select the ids.

However, there would be a limitation, because in this way it's not possible to specify at the same time a custom node_sets and a query dict containing a node set name, for example when the query is:
```python
{"$node_set": "Node12_L6_Y", "node_id": 1}
```

Comment (joni-herttuainen) [reply bluepysnap/edges/edges.py:49]: yeah, we can do something like
```python
sim.circuit.get({
    "$node_set": "Node12_L6_Y", 
    "$node_sets": sim.node_sets,  # probably better to have a more distinctive name like "$node_sets_instance" 
})
```

Comment (mgeplf) [reply bluepysnap/edges/edges.py:49]: is that what you were thinking @GianlucaFicarelli? I thought there might be a more clean solution.

Comment (GianlucaFicarelli) [reply bluepysnap/edges/edges.py:49]: When filtering by just a single node_set (from a node_sets that's not from the circuit, but from the simulation), I was thinking to:
```python
node_set = sim.node_sets["Node12_L6_Y"]
sim.circuit.get(node_set)
```
If the node_sets come from an external file (not circuit, not simulation), it could be:
```python
node_sets = NodeSets.from_file(...) # or NodeSets(...)
node_set = node_sets["Node12_L6_Y"]
sim.circuit.get(node_set)
```
We can still keep the usual shorter way (when the node_sets is circuit.node_sets):
```python
sim.circuit.get("Node12_L6_Y")
```

If a query is needed when using a different node_set at the same time, something like:
```python
node_set = sim.node_sets["Node12_L6_Y"]
sim.circuit.get({"$node_set": node_set, "morphology": "morph-B")})
```
The downside of this query is that it cannot be dumped as regular json, but maybe this is not desired or required.

About using different keys for node_set name and node_set instance, I'm not convinced that it would make the API easier, but it can be discussed.

File: bluepysnap/nodes/node_population.py
Comment (GianlucaFicarelli) [reply bluepysnap/nodes/node_population.py:274]: Yes this one. I don't have a strong opinion either, let's see if there are other opinions or known use cases.

File: bluepysnap/node_sets.py
Comment (GianlucaFicarelli) [reply bluepysnap/node_sets.py:70]: If we consider public the NodeSets class, I'd like if both snap and libsonata allow to init the object from str, path, dict.

libsonata already allows:
- `libsonata.NodeSets(<content>)`
- `libsonata.NodeSets.from_file(<path>)`

so it's missing only the init from a dict. Was it discussed before @mgeplf ?

If we want, we can still wrap around libsonata if we define in snap:
- `NodeSets.from_file()`
- `NodeSets.from_dict()`
- `NodeSets.from_string()`

or we can avoid one of them, if we define the init accordingly.

These methods probably don't have a lot of applications, but the main use case could be to build custom node_sets in memory (as I would like to do if it were possible to convert a query group to node_set, as I was trying to do in another PR).

Comment (mgeplf) [reply bluepysnap/node_sets.py:70]: > Was it discussed before @mgeplf ?

No; I'm a bit sad I did the `libsonata.NodeSets(<content>)`, so I think having it clear w/ `from_file()`, `from_string()` makes good sense.

Comment (joni-herttuainen) [reply bluepysnap/node_sets.py]: Added all three suggested methods

File: bluepysnap/node_sets.py
Comment (joni-herttuainen) [reply bluepysnap/node_sets.py]: thanks. Tunnel-visioned big time there.

File: bluepysnap/simulation.py
Comment (GianlucaFicarelli) [inline bluepysnap/simulation.py:129]: This wasn't changed in this PR, but it would be more consistent if we always return a NodeSet instance, instead of returning an empty dict when `node_sets_file` is empty.
What about of something like this?
```python
        return NodeSets.from_file(node_sets_file) if node_sets_file else NodeSets.from_string("{}")
```
or 
```python
        return NodeSets.from_file(node_sets_file) if node_sets_file else NodeSets.from_dict({})
```

File: doc/source/notebooks/03_node_properties.ipynb
Comment (GianlucaFicarelli) [inline doc/source/notebooks/03_node_properties.ipynb:191]: Just wondering, why images are now 640x480 instead of 432x288?

File: doc/source/notebooks/03_node_properties.ipynb
Comment (GianlucaFicarelli) [inline doc/source/notebooks/03_node_properties.ipynb:324]: This warning is new, is it possible to fix it? (here or in a different PR)

File: doc/source/notebooks/03_node_properties.ipynb
Comment (GianlucaFicarelli) [inline doc/source/notebooks/03_node_properties.ipynb:502]: If I'm not wrong, this image isn't present in the modified notebook

File: bluepysnap/simulation.py
Comment (GianlucaFicarelli) [reply bluepysnap/simulation.py:129]: Alternatively, we could return None, but it seems a bigger breaking change.

File: doc/source/notebooks/03_node_properties.ipynb
Comment (joni-herttuainen) [reply doc/source/notebooks/03_node_properties.ipynb:191]: Not sure. I mean the default seems to have been 640x480 for a while:
https://matplotlib.org/stable/users/prev_whats_new/dflt_style_changes.html#figure-size-font-size-and-screen-dpi

So the question is more like why was it 432x288?

File: doc/source/notebooks/03_node_properties.ipynb
Comment (joni-herttuainen) [reply doc/source/notebooks/03_node_properties.ipynb]: I added code to temporarily ignore it since I didn't find an immediate solution.

File: doc/source/notebooks/03_node_properties.ipynb
Comment (joni-herttuainen) [reply doc/source/notebooks/03_node_properties.ipynb:502]: Yeah, sometimes when I run it on a local (well, inside bb5) jupyter instance, it skips a cell. this is what happened here. Should be fixed now

**Comment by mgeplf** [comment]
Things look reasonable to me, but I haven't been following it as closely as @GianlucaFicarelli - IMO he should have the final say.

File: bluepysnap/simulation.py
Comment (GianlucaFicarelli) [reply bluepysnap/simulation.py:129]: Good, this is fixed in `Circuit.node_sets`, but it should be changed in `Simulation.node_sets` as well (and the changelog updated to mention Simulation)

Comment (joni-herttuainen) [reply bluepysnap/simulation.py]: 🤦‍♂️ , I remembered that when there is no simulation node set file defined, libsonata returns circuits node set path so this needed no fixing. OFC, both can be empty. 

Thanks > should be fixed now.

**Review by GianlucaFicarelli — COMMENTED** [review]
LGTM 👍 

Should we release 1.0.7 before merging and releasing 1.1.0?
Alternatively, the changelog should be updated (moving the entries from 1.0.7 to 1.1.0)

**Comment by joni-herttuainen** [comment]
I'm happy with merging and releasing any none-breaking changes first

**Comment by GianlucaFicarelli** [comment]
> I'm happy with merging and releasing any none-breaking changes first

I'd like to merge https://github.com/BlueBrain/snap/pull/220 once it's approved (since it's not a breaking change), release 1.0.7, then merge this PR and release 1.1.0. What do you think?

**Comment by joni-herttuainen** [comment]
Sounds good. I think #220 is good to go.

File: bluepysnap/nodes/node_population.py
Comment (GianlucaFicarelli) [reply bluepysnap/nodes/node_population.py]: Do you mind repeating the same benchmark with version 1.0.7, after merging or rebasing on master?

Comment (joni-herttuainen) [reply bluepysnap/nodes/node_population.py]: with `snap`:
```
# ROUNDS: 100

# Layer23Inhibitory #
- total time: 5.96 s
- per round (avg): 0.060s

# S1HL #
- total time: 2.55 s
- per round (avg): 0.025s

# mc2;VPL #
- total time: 0.66 s
- per round (avg): 0.007s

# mc2_Column #
- total time: 6.51 s
- per round (avg): 0.065s

--------------------------------
# ROUNDS: 1000

# Layer23Inhibitory #
- total time: 27.48 s
- per round (avg): 0.027s

# S1HL #
- total time: 19.42 s
- per round (avg): 0.019s

# mc2;VPL #
- total time: 6.56 s
- per round (avg): 0.007s

# mc2_Column #
- total time: 66.19 s
- per round (avg): 0.066s
```

with `libsonata`:
```
# ROUNDS: 100

# Layer23Inhibitory #
- total time: 3.39 s
- per round (avg): 0.034s

# S1HL #
- total time: 9.93 s
- per round (avg): 0.099s

# mc2;VPL #
- total time: 3.95 s
- per round (avg): 0.040s

# mc2_Column #
- total time: 7.02 s
- per round (avg): 0.070s

--------------------------------
# ROUNDS: 1000

# Layer23Inhibitory #
- total time: 26.13 s
- per round (avg): 0.026s

# S1HL #
- total time: 103.41 s
- per round (avg): 0.103s

# mc2;VPL #
- total time: 40.42 s
- per round (avg): 0.040s

# mc2_Column #
- total time: 71.51 s
- per round (avg): 0.072s
```

**Review by GianlucaFicarelli — APPROVED** [review]
LGTM

**Pull request merged**
